src/riscv/lib/src/jit/builder/control_flow_graph.rs

```diff
@@ -248,6 +276,83 @@ where
 
         step_updates
     }
+
+    /// Iterative DFS to detect cycles in the control flow graph.
+    /// Marks edges that close cycles as budget check locations.
+    pub fn find_budget_check_locations(&self) -> OutcomeMap<Option<BudgetCheckLoc<'_, T>>> {
+        let mut budget_checks = self.outcomes.map(|_, _| None);
+        let mut nodes = self.nodes.map(|_, _| NodeState::Unvisited);
+
+        let mut work_stack = self
+            .graph
+            .outgoing_outcomes(SourceInstrLoc::Entry)
+            .iter()
+            .filter_map(|outcome| {
+                self.outcomes[*outcome].to().as_internal().map(|&node_id| {
+                    (
+                        node_id,
+                        self.graph
+                            .outgoing_outcomes(SourceInstrLoc::Internal(node_id))
+                            .iter()
+                            .sorted_by_key(|child_outcome| {
+                                self.outcomes[**child_outcome].priority() as i32
+                            }),
+                    )
+                })
+            })
+            .collect_vec();
+
+        while let Some((node_id, mut child_outcomes)) = work_stack.pop() {
+            nodes[node_id] = NodeState::Visiting;
+
+            if let Some(child_outcome) = child_outcomes.next() {
+                // Push the current node back onto the stack to process remaining children later.
+                work_stack.push((node_id, child_outcomes));
+
+                // Process the child outcome.
+                match self.outcomes[*child_outcome].to() {
+                    TargetInstrLoc::Exit => {
+                        continue;
+                    }
+                    TargetInstrLoc::Internal(child_node_id) => match nodes[child_node_id] {
+                        NodeState::Visited => continue,
```

Comment:
it's not immediately obvious why `Visited` doesn't mean we have a cycle. Could you potentially rename this variant to something that explains it is guaranteed to exit/encounter budget check before recursing?